cirq/protocols/apply_mixture_protocol.py

```diff
@@ -0,0 +1,309 @@
+        mutating `target_tensor` it will return `target_tensor`. The caller is
+        responsible for checking if the result is `target_tensor`.
+
+        If the receiving object wrote its output over `out_buffer`, the
+        result will be `out_buffer`. The caller is responsible for
+        checking if the result is `out_buffer` (and e.g. swapping
+        the buffer for the target tensor before the next call).
+
+        Note that it is an error for the return object to be either of the
+        auxiliary buffers, and the method will raise an AssertionError if
+        this contract is violated.
+
+        The receiving object may also write its output over a new buffer
+        that it created, in which case that new array is returned.
+
+    Raises:
+        TypeError: `val` doesn't have a mixture and `default` wasn't specified.
+        ValueError: Different left and right shapes of `args.target_tensor`
+            selected by `left_axes` and `right_axes` or `qid_shape(val)` doesn't
+            equal the left and right shapes.
+        AssertionError: `_apply_mixture_` returned an auxiliary buffer.
+    """
+
+    # Verify that val has the same qid shape as the selected axes of the density
+    # matrix tensor.
+    val_qid_shape = qid_shape_protocol.qid_shape(val,
+                                                 (2,) * len(args.left_axes))
+    left_shape = tuple(args.target_tensor.shape[i] for i in args.left_axes)
+    right_shape = tuple(args.target_tensor.shape[i] for i in args.right_axes)
+    if left_shape != right_shape:
+        raise ValueError('Invalid target_tensor shape or selected axes. '
+                         'The selected left and right shape of target_tensor '
+                         'are not equal. Got {!r} and {!r}.'.format(
+                             left_shape, right_shape))
+    if val_qid_shape != left_shape:
+        raise ValueError('Invalid mixture qid shape is not equal to the '
+                         'selected left and right shape of target_tensor. '
+                         'Got {!r} but expected {!r}.'.format(
+                             val_qid_shape, left_shape))
+
+    # Check if the specialized method is present.
+    func = getattr(val, '_apply_mixture_', None)
+    if func is not None:
+        result = func(args)
+        if result is not NotImplemented and result is not None:
+
+            def err_str(buf_num_str):
+                return ("Object of type '{}' returned a result object equal to "
+                        "auxiliary_buffer{}. This type violates the contract "
+                        "that appears in apply_mixture's documentation.".format(
+                            type(val), buf_num_str))
+
+            assert result is not args.auxiliary_buffer0, err_str('0')
+            assert result is not args.auxiliary_buffer1, err_str('1')
+            return result
+
+    # Possibly use `apply_unitary`.
+    result = _apply_unitary(val, args)
```

Comment:
You also need a decompose based strategy.

Because the decompose here will be redundant with the decompose inside apply unitary, we should consider inlining some of the apply unitary logic. Specifically, two of the strategies will be "look for `_apply_unitary_` and try to use it" as well as "look for `_unitary_` and try to use it".

Reply:
the protocols.apply_unitary attempt here tries to use the decompose funcitonality. Am I missing something here ?